Ctor 1320 enhance sc prometheus gateway apiv1 to prometheus gateway events apiv2 and prometheus gateway metrics apiv2

omercier opened the pull request:
## Description

**PLEASE MAKE SURE THAT THE BRANCH PR INCLUDES JIRA TICKET ID** (_for centreon-internal_)

Please include a short resume of the changes and what is the purpose of PR. Any relevant information should be added to help:
* **QA Team** (Quality Assurance) with tests.
* **reviewers** to understand what are the stakes of the pull request.

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [ ] 24.04.x
- [ ] 24.10.x
- [ ] master

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

## Checklist

#### Community contributors & Centreon team

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

omercier committed:
wip: things to improve?

omercier committed:
feat(prometheus-metrics-v2): new connector for metrics

omercier committed:
fix CTOR-1626

omercier committed:
enh(libs): log an error when log_level is ignored because of its type

omercier committed:
feat(prometheus-events-v2): new connector for events

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-events-apiv2.lua:
```diff
@@ -0,0 +1,378 @@
+    return false
+  end
+
+  return string.gsub(string, '[^a-zA-Z0-9_:]', '_')
```
need to use the the metric_pattern param instead of a hardcoded pattern

omercier replied:
I'll look for that 👍🏻

omercier replied:
Fixed here https://github.com/centreon/centreon-stream-connector-scripts/pull/279/commits/76eab9f8f000eb1958378c06306ed5ae4f572c41

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-events-apiv2.lua:
```diff
@@ -0,0 +1,378 @@
+
+-- Libraries
+local curl      = require "cURL"
+local base64    = require("base64")
```
i don't know this one. I'm using b64 from the mime part of the luasocket lib since it is already a dependency

omercier replied:
I took it from the original prometheus connector. I'll switch.

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-events-apiv2.lua:
```diff
@@ -0,0 +1,378 @@
+    data = data .. name .. '{label="monitoring_status", host="' .. payload.prom_hname .. '", service="' .. payload.prom_sdesc .. '"} ' .. payload.state .. '\n'
+  else
+    data = data .. name .. '{label="monitoring_status", host="' .. payload.prom_hname .. '", service="' .. payload.prom_sdesc .. '", ' ..  payload.hostgroupsLabel .. '} ' .. payload.state .. '\n'
+  end
```
should considere having that in the "format" fucntions

omercier replied:
Fixed there https://github.com/centreon/centreon-stream-connector-scripts/pull/279/commits/5d9cedb114099eb2ee6d5348e4d3e698bdc278b8

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-events-apiv2.lua:
```diff
@@ -0,0 +1,378 @@
+    self.events = {}
+    retval = true
+  else
+    self.sc_logger:error("EventQueue:send_data: HTTP POST request FAILED, return code is " .. httpResponseCode .. " message is:\n\"" .. httpResponseBody .. "\n\"\n")
```
```suggestion
    self.sc_logger:error("EventQueue:send_data: HTTP POST request FAILED, return code is " .. httpResponseCode .. " message is:\n\"" .. tostring(httpResponseBody) .. "\n\"\n")
```

just to be sure :D

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-events-apiv2.lua:
```diff
@@ -0,0 +1,378 @@
+  self.sc_params.params.prometheus_url         = params.prometheus_url or "http://127.0.0.1:9091"
+  self.sc_params.params.http_timeout           = params.http_timeout or 30
+  self.sc_params.params.prometheus_gateway_job = params.prometheus_gateway_job or "monitoring"
+  
```
```suggestion
  -- force max_buffer_size to 1 because we each service is sent to its own url
  self.sc_params.params.max_buffer_size = 1
  
```

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-events-apiv2.lua:
```diff
@@ -0,0 +1,378 @@
+    data = data .. name .. '{label="monitoring_status", host="' .. payload.prom_hname .. '", service="' .. payload.prom_sdesc .. '", ' ..  payload.hostgroupsLabel .. '} ' .. payload.state .. '\n'
+  end
+
+
```
missing the send_data_test and log_curl_command features

omercier replied:
fixed here https://github.com/centreon/centreon-stream-connector-scripts/pull/279/commits/b386ac65b9d1c10786fe7a9be34ca8fb570f334e

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-events-apiv2.lua:
```diff
@@ -0,0 +1,378 @@
+    self.sc_logger:error("the body request " .. data)
+  end
+
+  -- and update the timestamp
+  self.__internal_ts_last_flush = os.time()
```
```suggestion
```

this is done by the sc_flush library already https://github.com/centreon/centreon-stream-connector-scripts/blob/8278b591981357b9f17d798c4a8ac77ec6d83961/modules/centreon-stream-connectors-lib/sc_flush.lua#L107

tanguyvda commented on modules/centreon-stream-connectors-lib/sc_params.lua:
```diff
@@ -1011,6 +1011,11 @@ function ScParams:check_params()
   self.params.logfile = self.common:ifnil_or_empty(self.params.logfile, "/var/log/centreon-broker/stream-connector.log")
+
+  if type(self.params.log_level) ~= "number" then
+    self.logger:error("[sc_params:check_params]: log_level parameter given as a " .. type(self.params.log_level) .. " (" .. self.params.log_level .. ") instead of a number. Ignored.")
```
this is the death of fun 😈

omercier committed:
force max_buffer_size to 1 because we each service is sent to its own url

Co-authored-by: tcharles <tcharles@centreon.com>

omercier committed:
convert httpResponseBody to string before logging it

Co-authored-by: tcharles <tcharles@centreon.com>

omercier committed:
remove useless __internal_ts_last_flush update

Co-authored-by: tcharles <tcharles@centreon.com>

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+
+-- Libraries
+local curl        = require "cURL"
+local base64      = require("base64")
```
same story about this dependency

omercier replied:
Done here https://github.com/centreon/centreon-stream-connector-scripts/pull/279/commits/ab2a677c29943d37415fb088c52e364ca6321f05

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+    return false
+  end
+
+  return string.gsub(string, '[^a-zA-Z0-9_:]', '_')
```
same story about de hardocded thingy

omercier replied:
fixed here https://github.com/centreon/centreon-stream-connector-scripts/pull/279/commits/76eab9f8f000eb1958378c06306ed5ae4f572c41

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+    A = 'amperes',
+    K = 'kelvins',
+    ratio = 'ratios',
+    degres = 'celsius'
```
```suggestion
    ["%"] = 'ratios',
    ["°"] = 'celsius'
```

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+    if (unitMapping[unit] ~= nil) then
+      unit = unitMapping[unit]
+    end
+  end
```
```suggestion
if unitMapping[unit] then
  unit = unitMapping[unit]
end
```

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+  local type  = self:get_metric_type(metric)
+  local unit  = unit_mapping(metric.uom)
+  local label = metric.metric_name
+  local name  = self:create_metric_name(label, unit)
```
I forgot it in my clickhouse sc and it has been forgotten in the influxdb1 sc. This metric_name doesn't include the instance and subinstances part of the metric name. Which is not problematic but you either need to include them in the label or add them as "metadata" of your event like you do for the service name and host name.

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+  self.sc_logger:debug("[EventQueue:format_metric]: end real format metric ")
+end
+
+--------------------------------------------------------------------------------
+-- ifnumber_not_nan: [for Prometheus] check if a number is a number (and not a NaN)
+-- @param {number} number, the number to check
+-- @return {boolean}
+--------------------------------------------------------------------------------
+local function ifnumber_not_nan (number)
+  if (number ~= number) then
+    return false
+  elseif (type(number) ~= 'number') then
+    return false
+  else
+    return true
+  end
+end
```
```suggestion
--------------------------------------------------------------------------------
-- ifnumber_not_nan:  check if a number is a number (and not a NaN)
-- @param {number} number, the number to check
-- @return {boolean}
--------------------------------------------------------------------------------
local function is_number_and_not_a_NaN (number)
  if (number ~= number) then
    return false
  end
  
  if (type(number) ~= "number") then
    return false
  end

  return true
end
```

- This function can be useful outside of this stream connector so it must be put in the sc_common lib instead (and documented 😈 
- the suggestion is how I would write it nowadays (i've not updated the comment but i've changed the name of the function because the best practice when creating a function that returns a boolean is to start it with "is_" something...) 
-

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+    self.sc_logger:error("EventQueue:send_data: HTTP POST request FAILED, return code is " .. httpResponseCode .. " message is:\n\"" .. httpResponseBody .. "\n\"\n")
+    self.sc_logger:error("the body request " .. httpPostData)
+  end
+
+  -- and update the timestamp
+  self.__internal_ts_last_flush = os.time()
+
```
```suggestion

```

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+  self.sc_logger:debug("[EventQueue:send_data]: Starting to send data")
+  local httpPostData = payload.payload
+  local httpResponseBody = ""
+
```
missing the send_data_test and log_curl_command feature

omercier replied:
fixed here https://github.com/centreon/centreon-stream-connector-scripts/pull/279/commits/b386ac65b9d1c10786fe7a9be34ca8fb570f334e

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+    self.events = {}
+    retval = true
+  else
+    self.sc_logger:error("EventQueue:send_data: HTTP POST request FAILED, return code is " .. httpResponseCode .. " message is:\n\"" .. httpResponseBody .. "\n\"\n")
```
```suggestion
    self.sc_logger:error("EventQueue:send_data: HTTP POST request FAILED, return code is " .. httpResponseCode .. " message is:\n\"" .. tostring(httpResponseBody) .. "\n\"\n")
```

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+--------------------------------------------------------------------------------
+function EventQueue:get_metric_type (perfdata)
+  local metricType = nil;
+  if (ifnumber_not_nan(perfdata.max)) then
```
will need to change this function call to use the proper one

omercier replied:
done here https://github.com/centreon/centreon-stream-connector-scripts/pull/279/commits/76eab9f8f000eb1958378c06306ed5ae4f572c41

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+function EventQueue:get_metric_type (perfdata)
+  local metricType = nil;
+  if (ifnumber_not_nan(perfdata.max)) then
+    metricType = 'gauge'
+  else 
+    metricType = 'counter'
+  end
+  
+  return metricType
```
```suggestion
    return "gauge"
  end
  
  return "counter"
```

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+-- @return {string} metricType, the type of the metric
+--------------------------------------------------------------------------------
+function EventQueue:get_metric_type (perfdata)
+  local metricType = nil;
```
```suggestion
```

tanguyvda commented on centreon-certified/prometheus/prometheus-pushgateway-metrics-apiv2.lua:
```diff
@@ -0,0 +1,613 @@
+  local hname = self.sc_event.event.cache.host.name
+
+  if (unit ~= '') then
+    if (self.enable_extended_metric_name == 0) then
+      name = label .. '_' .. unit
+    else
+      name = hname .. '_' .. sdesc .. ':' .. label .. '_' .. unit
+    end
+  else
+    if (self.enable_extended_metric_name == 0) then
```
this parameter doesn't exist. Must be added in the EventQueue constructor

omercier replied:
fixed there https://github.com/centreon/centreon-stream-connector-scripts/pull/279/commits/6e5bfd672680c331dbd2bd5cc87dfc84c2debd31

tanguyvda reviewed (changes requested):
back in the days, I've made a lot of camelCase syntaxes. Need to switch them to snake_case

omercier committed:
enh(lib): use mime.b64

omercier committed:
enh: move formatting of payload where it belongs

omercier committed:
Apply suggestions from code review by Tanguy

Co-authored-by: tcharles <tcharles@centreon.com>

omercier committed:
fix: default value for log_level was empty string instead of number

omercier committed:
end: use existing generic parameters to format metrics instead of hard coded

omercier committed:
enh: handle enable_extended_metric_name param

omercier committed:
enh: add log_curl_command and send_data debug

omercier commented:
> back in the days, I've made a lot of camelCase syntaxes. Need to switch them to snake_case

This comment does not point anywhere in the code. Are you asking me to refactor all camelCase variables and functions?

omercier committed:
enh: various enhancements

- metrics naming looks ok: no unit redundancy in names
- instances and subinstances handled
- removed hosts events since theyre not relevant
- the headers were not displayed by log_curl_command

omercier committed:
fix: fixed some silly mistakes

omercier committed:
enh: partially convert camelCase to snake_case

omercier committed:
finish removing camelCase

omercier committed:
enh: improve self-documentation

omercier committed:
fix: handle host names with special characters + other

- rename url parameter to match specs
- enable smart sleep for events too
- rewrite end of write() function

omercier committed:
enh: add get_hostgroup_alias

omercier committed:
enh sc_broker documentation

omercier committed:
enh sc_flush doc

omercier committed:
enh: quick doc enhancement

omercier committed:
enh: improve documentaiton of sc_event²

omercier committed:
enh prometheus events sc
